<a href="https://colab.research.google.com/github/jamestheengineer/think-stats/blob/main/python/chap01ex.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
# Only do this once per VM, otherwise you'll get multiple clones and nested directories
#!git clone https://github.com/AllenDowney/ThinkStats2.git
%cd ThinkStats2/code

/content/ThinkStats2/code


# Examples and Exercises from Think Stats, 2nd Edition

http://thinkstats2.com

Copyright 2016 Allen B. Downey

MIT License: https://opensource.org/licenses/MIT


In [4]:
from __future__ import print_function, division

import nsfg

## Examples from Chapter 1

Read NSFG data into a Pandas DataFrame.

In [5]:
preg = nsfg.ReadFemPreg()
preg.head()

caseid  pregordr  howpreg_n  howpreg_p  ...  secu_p  sest  cmintvw  totalwgt_lb
0       1         1        NaN        NaN  ...       2     9      NaN       8.8125
1       1         2        NaN        NaN  ...       2     9      NaN       7.8750
2       2         1        NaN        NaN  ...       2    12      NaN       9.1250
3       2         2        NaN        NaN  ...       2    12      NaN       7.0000
4       2         3        NaN        NaN  ...       2    12      NaN       6.1875

[5 rows x 244 columns]

Print the column names.

In [6]:
preg.columns

Index(['caseid', 'pregordr', 'howpreg_n', 'howpreg_p', 'moscurrp', 'nowprgdk',
       'pregend1', 'pregend2', 'nbrnaliv', 'multbrth',
       ...
       'laborfor_i', 'religion_i', 'metro_i', 'basewgt', 'adj_mod_basewgt',
       'finalwgt', 'secu_p', 'sest', 'cmintvw', 'totalwgt_lb'],
      dtype='object', length=244)

Select a single column name.

In [7]:
preg.columns[1]

'pregordr'

Select a column and check what type it is.

In [8]:
pregordr = preg['pregordr']
type(pregordr)

pandas.core.series.Series

Print a column.

In [9]:
pregordr

0        1
1        2
2        1
3        2
4        3
        ..
13588    1
13589    2
13590    3
13591    4
13592    5
Name: pregordr, Length: 13593, dtype: int64

Select a single element from a column.

In [10]:
pregordr[0]

1

Select a slice from a column.

In [11]:
pregordr[2:5]

2    1
3    2
4    3
Name: pregordr, dtype: int64

Select a column using dot notation.

In [12]:
pregordr = preg.pregordr

Count the number of times each value occurs.

In [13]:
preg.outcome.value_counts().sort_index()

1    9148
2    1862
3     120
4    1921
5     190
6     352
Name: outcome, dtype: int64

Check the values of another variable.

In [14]:
preg.birthwgt_lb.value_counts().sort_index()

0.0        8
1.0       40
2.0       53
3.0       98
4.0      229
5.0      697
6.0     2223
7.0     3049
8.0     1889
9.0      623
10.0     132
11.0      26
12.0      10
13.0       3
14.0       3
15.0       1
Name: birthwgt_lb, dtype: int64

Make a dictionary that maps from each respondent's `caseid` to a list of indices into the pregnancy `DataFrame`.  Use it to select the pregnancy outcomes for a single respondent.

In [15]:
caseid = 10229
preg_map = nsfg.MakePregMap(preg)
indices = preg_map[caseid]
preg.outcome[indices].values

array([4, 4, 4, 4, 4, 4, 1])

## Exercises

Select the `birthord` column, print the value counts, and compare to results published in the [codebook](http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=A&subSec=8016&srtLabel=611933)

In [16]:
# Solution goes here

We can also use `isnull` to count the number of nans.

In [17]:
preg.birthord.isnull().sum()

4445

Select the `prglngth` column, print the value counts, and compare to results published in the [codebook](http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=A&subSec=8016&srtLabel=611931)

In [18]:
# Solution goes here

To compute the mean of a column, you can invoke the `mean` method on a Series.  For example, here is the mean birthweight in pounds:

In [19]:
preg.totalwgt_lb.mean()

7.265628457623368

Create a new column named <tt>totalwgt_kg</tt> that contains birth weight in kilograms.  Compute its mean.  Remember that when you create a new column, you have to use dictionary syntax, not dot notation.

In [20]:
# Solution goes here

`nsfg.py` also provides `ReadFemResp`, which reads the female respondents file and returns a `DataFrame`:

In [21]:
resp = nsfg.ReadFemResp()

`DataFrame` provides a method `head` that displays the first five rows:

In [22]:
resp.head()

caseid  rscrinf  rdormres  ...  cmlstyr  screentime   intvlngth
0    2298        1         5  ...     1222    18:26:36  110.492667
1    5012        1         5  ...     1221    16:30:59   64.294000
2   11586        1         5  ...     1222    18:19:09   75.149167
3    6794        5         5  ...     1222    15:54:43   28.642833
4     616        1         5  ...     1221    14:19:44   69.502667

[5 rows x 3087 columns]

Select the `age_r` column from `resp` and print the value counts.  How old are the youngest and oldest respondents?

In [23]:
# Solution goes here

We can use the `caseid` to match up rows from `resp` and `preg`.  For example, we can select the row from `resp` for `caseid` 2298 like this:

In [24]:
resp[resp.caseid==2298]

caseid  rscrinf  rdormres  ...  cmlstyr  screentime   intvlngth
0    2298        1         5  ...     1222    18:26:36  110.492667

[1 rows x 3087 columns]

And we can get the corresponding rows from `preg` like this:

In [25]:
preg[preg.caseid==2298]

caseid  pregordr  howpreg_n  ...  sest  cmintvw  totalwgt_lb
2610    2298         1        NaN  ...    18      NaN       6.8750
2611    2298         2        NaN  ...    18      NaN       5.5000
2612    2298         3        NaN  ...    18      NaN       4.1875
2613    2298         4        NaN  ...    18      NaN       6.8750

[4 rows x 244 columns]

How old is the respondent with `caseid` 1?

In [26]:
# Solution goes here

What are the pregnancy lengths for the respondent with `caseid` 2298?

In [27]:
# Solution goes here

What was the birthweight of the first baby born to the respondent with `caseid` 5012?

In [28]:
# Solution goes here